### PR TITLE
Feat: Add feeFactory

### LINKED ISSUE #25: As a User I want to define fees for the collection
**Acceptance criteria:**
- All of the fee types should be supported:
  - Fixed fee
  - Royalty fee
  - Royalty fee with a fallback fee
- there is a documentation with examples for all of the supported types

example use of the SDK:
``` typescript
const fee = FeeFactory.RoyaltyFee(
  collectorAccountId: "0.0.123",
  nominator: 1,
  denominator: 100,
  fallbackFee: fixedFee,
  allCollectorsAreExempt: true
);
```

### PR DESCRIPTION
**Pull request by StanislawMazowieckiAriane**
**Description**:

**Related issue(s)**:

Fixes #25 

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: src/functions/createFixedFeeFunction.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    fixedFee.setHbarAmount(Hbar.fromString(hbarAmount.toString()));
+  }
+
+  if (amount) {
+    fixedFee.setAmount(amount);
+  }
+
+  if (denominatingTokenId) {
+    fixedFee.setDenominatingTokenId(denominatingTokenId);
+  }
```
Comment (BartoszSolkaArianelabs) [inline src/functions/createFixedFeeFunction.ts:25]: `hbarAmount`, `amount` and `denominatingTokenId` are related to each other.

Either `hbarAmount` is set and (`amount` and `denominatingTokenId`) are not set
**OR**
(`amount` and `denominatingTokenId`) are set and `hbarAmount` is not set

File: src/types/validateProps.ts [hunk 2]
```diff
@@ -22,4 +22,14 @@ export type validateCreateCollectionProps = {
   collectionSymbol?: string;
   treasuryAccountPrivateKey?: string;
   treasuryAccount?: string;
+  customFees?: CustomFee[];
+};
+
+export type validateFixedFeeFunction = {
```
Comment (se7enarianelabs) [inline src/types/validateProps.ts:28]: It's not a function rename, this to something like: `fixedFeeValidationProps`

File: src/types/validateProps.ts [hunk 3]
```diff
@@ -22,4 +22,14 @@ export type validateCreateCollectionProps = {
...
+  collectorAccountId?: string;
+};
+
+export type validateRoyaltyFeeFunction = validateFixedFeeFunction & {
```
Comment (se7enarianelabs) [inline src/types/validateProps.ts:32]: It's not a function rename, this to something like: royaltyFeeValidationProps

**Review by se7enarianelabs — CHANGES_REQUESTED** [review]

File: test/e2e/feeFactoryE2E.test.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import { feeFactoryInstance, nftSDK } from './e2eConsts';
+import { beforeEach } from 'node:test';
+import { HederaNFTSDK } from '../../src/HederaNFTSDK';
+import { myAccountId, myPrivateKey, mySecondAccountId } from '../__mocks__/consts';
+import { dictionary } from '../../src/utils/constants/dictionary';
+
+beforeEach(async () => {
+  new HederaNFTSDK(myAccountId, myPrivateKey);
+});
```
[the remark this answers is not in the document]
Comment (StanislawMazowieckiAriane) [reply test/e2e/feeFactoryE2E.test.ts:9]: But I'm creating collections in this test. For this, I need to be logged in

[the remark this answers is not in the document]
Comment (BartoszSolkaArianelabs) [reply test/e2e/feeFactoryE2E.test.ts:9]: `nftSDK` is already created with these keys in test/e2e/e2eConsts.ts

[the remark this answers is not in the document]
Comment (StanislawMazowieckiAriane) [reply test/e2e/feeFactoryE2E.test.ts:9]: You're right, sorry

File: src/types/validateProps.ts (hunk 3 above)
Comment (StanislawMazowieckiAriane) [reply src/types/validateProps.ts:32]: Done

File: src/types/validateProps.ts (hunk 2 above)
Comment (StanislawMazowieckiAriane) [reply src/types/validateProps.ts:28]: Done

File: src/functions/createFixedFeeFunction.ts (hunk 1 above)
Comment (StanislawMazowieckiAriane) [reply src/functions/createFixedFeeFunction.ts:25]: Done

File: src/utils/validateProps.ts [hunk 5]
```diff
@@ -30,29 +30,39 @@ export const validatePropsForCreateCollection = (props: validateCreateCollection
...
 
-export const validatePropsForRoyaltyFeeFunction = (props: validateRoyaltyFeeFunction) => {
+export const validatePropsForRoyaltyFeeFunction = (props: royaltyFeeValidationProps) => {
   validateCollectorAccountId(props);
   validateNumerator(props);
   validateDenominator(props);
 };
 
-const validateNumerator = (props: validateRoyaltyFeeFunction) => {
+const validateHbarAmountAmountAndDenominatingToken = (props: fixedFeeValidationProps) => {
```
Comment (BartoszSolkaArianelabs) [inline src/utils/validateProps.ts:44]: please write unit tests for all possible combinations

(`false` - not set, `true` - set)

| hbarAmount | amount | denominatingTokenId | expectedResult |
|------------|--------|---------------------|----------------|
| false          | false      | false                   | error          |
| false          | false      | true                   | error          |
| false          | true      | false                   | error          |
| false          | true      | true                   | valid          |
| true          | false      | false                   | valid          |
| true          | false      | true                   | error          |
| true          | true      | false                   | error          |
| true          | true      | true                   | error          |

File: src/utils/validateProps.ts [hunk 6]
```diff
@@ -30,29 +30,39 @@ export const validatePropsForCreateCollection = (props: validateCreateCollection
...
+  if (
+    (props.hbarAmount && (props.amount || props.denominatingTokenId)) ||
+    (!props.hbarAmount && (!props.amount || !props.denominatingTokenId))
+  ) {
+    throw new Error(dictionary.createCollection.hbarAmountAmountAndDenominatingToken);
```
Comment (BartoszSolkaArianelabs) [inline src/utils/validateProps.ts:49]: `hbarAmountAmountAndDenominatingToken` can be renamed to `hbarAmountOrAmountAndDenominatingToken`

**Review by BartoszSolkaArianelabs — CHANGES_REQUESTED** [review]

File: src/utils/validateProps.ts (hunk 5 above)
Comment (StanislawMazowieckiAriane) [reply src/utils/validateProps.ts]: Done

File: src/utils/validateProps.ts (hunk 6 above)
Comment (StanislawMazowieckiAriane) [reply src/utils/validateProps.ts:49]: Done

**Review by se7enarianelabs — APPROVED** [review]

**Pull request merged**